File: Com-lib/UnityRobot_CommLib/communicator.cpp
#include "communicator.h"
#include "IPresentationProtocol.hpp"
#include "IDataLink.hpp"
#include "RobotLogger.h"
namespace UnityRobot
{

Communicator::Communicator(Networking::IPresentationProtocol& prot, Networking::IDataLink& data)
	: m_protocol(prot), m_datalink(data)
{
	if (!m_datalink.Connected())
	{
		LogCritError("Datalink provided to communicator is not connected");
		throw std::invalid_argument("Datalink is not connected");
	}
}

bool Communicator::sendCommand(const Communication::Message& m) const
{
	return m_datalink.Connected() ? m_datalink.SendData(m_protocol.MessageToBinaryData(m)) : false;
}

Networking::IDataLink& Communicator::getDatalink() const
{
	return m_datalink;
}

Networking::IPresentationProtocol& Communicator::getPresentationProtocol() const
{
	return m_protocol;
}

}

File: networking/networking/networking/Communicator.cs
using Communication;
using System;

namespace Networking
{
    public class Communicator : IMessageSender, IDisposable
    {
        private IPresentationProtocol _pp;
        private IDataLink _datalink;

        public Communicator(IDataLink dataLink, IPresentationProtocol pP)
        {
            if (pP == null)
            {
                throw new System.ArgumentException("Parameter cannot be null", "pP");
            }

            if (dataLink == null)
            {
                throw new System.ArgumentException("Parameter cannot be null", "dataLink");
            }

            _pp = pP;
            _datalink = dataLink;
        }

        ~Communicator()
        {
            Dispose();
        }

        public IDataLink GetDataLink()
        {
            return _datalink;
        }

        public IPresentationProtocol GetPresentationProtocol()
        {
            return _pp;
        }

        public bool SendCommand(Message message)
        {
            bool success = false;

            byte[] values = _pp.MessageToBinaryData(message);

            if(values.Length > 0)
            {
                if (_datalink.Connected())
                {
                    success = _datalink.SendData(values);
                }
            }

            return success;
        }

        public void Dispose()
        {
            _datalink.Dispose();
        }
    }
}
